pwiz_tools/Skyline/FileUI/PeptideSearch/HardklorSearchControl.cs

```diff
@@ -0,0 +1,153 @@
+using System.Windows.Forms;
+using pwiz.Common.SystemUtil;
+using pwiz.Skyline.Model;
+using pwiz.Skyline.Properties;
+using pwiz.Skyline.Util.Extensions;
+using pwiz.Skyline.Model.DdaSearch;
+
+namespace pwiz.Skyline.FileUI.PeptideSearch
+{
+    public class HardklorSearchControl : SearchControl
+    {
+        private ImportPeptideSearch ImportPeptideSearch;
+        private ImportPeptideSearch.HardklorSettings _settings;
+        private int _totalSteps;
+        private int _currentStep;
+        HardklorSearchEngine FeatureFinder => ImportPeptideSearch.SearchEngine as HardklorSearchEngine;
+
+
+        public HardklorSearchControl(ImportPeptideSearch importPeptideSearch)
+        {
+            ImportPeptideSearch = importPeptideSearch;
+        }
+
+        // Steps, in parallel, for n files:
+        //    Threads 0 through (n-1): per file, convert to mzML, then do HardKlor then Bullseye
+        //    Thread n: Perform alignments on all mzMLs as they become available
+        // Finally combine the results across replicates
+
+        private bool Search(ImportPeptideSearch.HardklorSettings settings, CancellationTokenSource token)
+        {
+            _settings = settings;
+            ParallelRunnerProgressControl multiProgressControl = null;
+            try
+            {
+                Invoke(new MethodInvoker(() =>
+                {
+                    multiProgressControl = new ParallelRunnerProgressControl(this);
+                    multiProgressControl.Dock = DockStyle.Fill;
+                    progressSplitContainer.Panel1.Controls.Add(multiProgressControl);
+                }));
+
+                FeatureFinder.Generate(multiProgressControl, this, token.Token);
+            }
+            catch (OperationCanceledException e)
+            {
+                UpdateProgress(_status.ChangeWarningMessage(e.InnerException?.Message ?? e.Message));
+                return false;
+            }
+            catch (Exception e)
+            {
+                UpdateProgress(_status.ChangeErrorException(e));
+                return false;
+            }
+            finally
+            {
+                Invoke(new MethodInvoker(() =>
+                {
+                    progressSplitContainer.Panel1Collapsed = true;
+                    progressSplitContainer.Panel1.Controls.Clear();
+                    multiProgressControl?.Dispose();
+                }));
+            }
+
+            return !token.IsCancellationRequested;
+        }
+
+        public override UpdateProgressResponse UpdateProgress(IProgressStatus status)
+        {
+            if (IsCanceled)
+            {
+                return UpdateProgressResponse.cancel;
+            }
+
+            lock (progressBar)
+            {
+                _status = _status.ChangeMessage(status.Message).ChangePercentComplete((100 * _currentStep++) / _totalSteps);
+                Invoke(new MethodInvoker(() => UpdateSearchEngineProgress(_status)));
```

Comment:
I don't love seeing Invoke() inside a lock {}. Invoke is inherently single-threaded because it blocks until the code can be run on the UI thread, which is a single thread. Maybe the lock should go only around the update to _status and it should be on "this" because it is inherently about keeping _currentStep, _totalSteps, and _status in synch. If those are in synch, I don't think you need any further synchronization on UpdateSearchEngineProgress().

In fact, I would generally recommend against forcing an update on the UI thread for every change to _status. Instead, you want updates to happen based on a timer that fires on the UI thread and checks the current _status whatever it is. Status updates are about letting the user know about progress, which the user only needs every so often, e.g. 200 ms. I have seen real performance issues that came down to code like this where every status update forces a UI update.

Even if it is not a problem here, writing status update code like this forces the programmer to consider how often they need to update status to balance user need to see some progress and the performance consideration that UI updates can be much slower than the processing they are reporting on.

Reply:
Thanks, I'll have to think about that. I was following a pattern from elsewhere in our code.

FWIW these status updates don't happen very often - just at the end of each completed process (msconvert, hardklor etc)

Reply:
Can't it just use BeginInvoke instead?

Reply:
> Can't it just use BeginInvoke instead?

I'll give that a try, thanks.

Reply:
If you are confident about the timing issue, then I think you can just move the Invoke outside of the lock. I don't see why it needs to be inside the lock.

The problem with putting it in the lock is that if the UI thread somehow triggers this code and stops at the lock, and while a background thread is blocked on the Invoke() then you get a deadlock.

If you feel everything else after your Invoke call can happen without waiting for it to complete, you can use BeginInvoke() here, but that also seems to be to indicate the Invoke() doesn't need to be in the lock {}. i.e. it means that the what happens in BeginInvoke() doesn't need to happen in the lock. Does the queueing of this work need to happen in the lock? I don't see why.

Reply:
I would use lock(this) and have it only contain the update to _status, which clearly involved multiple member variables you are trying to keep in sync. The update to the UI could happen at any time as long as the _status is coherent.

And, I would still say that in general status UI should be implemented as a timer checking for status updates rather than pushing the UI in real-time from a background thread. But, point taken that this may be good enough for the processing in question.

Reply:
OK, I'll switch the lock to (this)